examples/wmt/input_pipeline.py

```diff
@@ -0,0 +1,255 @@
+        bucket_length * 16
+    ]
+    bucket_batch_sizes = [
+        batch_size * 4, batch_size * 2, batch_size,
+        batch_size // 2, batch_size // 4, batch_size // 8,
+        batch_size // 16
+    ]
+    # allow for different evaluation max-length bucket and batchsize
+    if not training:
+      max_eval_length = max_eval_length or bucket_length * 32
+      bucket_boundaries[-1] = max_eval_length
+      bucket_batch_sizes[-1] = batch_size // max_eval_length
+    # We will pad to boundaries which pads to bucket_boundary - 1: add 1 here.
+    bucket_boundaries = [b + 1 for b in bucket_boundaries]
+    # Make batch sizes divisible by n_devices.
+    bucket_batch_sizes = [
+        max(b // n_devices, 1) * n_devices for b in bucket_batch_sizes
+    ]
+    buckets = (bucket_boundaries, bucket_batch_sizes)
+
+  logging.info('Bucketing with buckets %s.', str(buckets))
+
+  def example_length(example):
+    """The length function used by bucket_by_sequence_length to bucket."""
+    return tf.maximum(tf.shape(example['inputs'])[0],
+                      tf.shape(example['targets'])[0])
+
+  boundaries, batch_sizes = buckets
+  # bucket_by_sequence_length expects a final dummy 1 batch_size
+  batch_sizes.append(1)
+  dataset = dataset.apply(
+      tf.data.experimental.bucket_by_sequence_length(
+          example_length,
+          boundaries,
+          batch_sizes,
+          pad_to_bucket_boundary=True,
+          drop_remainder=drop_remainder))
+  return dataset
+
+
+def preprocess_wmt_data(dataset,
+                        training,
+                        n_devices,
+                        shuffle_buffer_size=1024,
+                        max_target_length=512,
+                        max_eval_target_length=2048,
+                        batch_size=256,
+                        buckets=None,
+                        drop_remainder=True,
+                        prefetch_size=AUTOTUNE):
+  """Shuffle and batch the given dataset."""
+
+  def length_filter(max_len):
+    def filter_fn(x):
+      source, target = x['inputs'], x['targets']
+      l = tf.maximum(tf.shape(source)[0], tf.shape(target)[0])
+      return tf.less(l, max_len + 1)
+    return filter_fn
+
+  if max_target_length > 0 and training:
+    dataset = dataset.filter(length_filter(max_target_length))
+
+  if max_eval_target_length > 0 and not training:
+    dataset = dataset.filter(length_filter(max_eval_target_length))
+
+  if training:
+    dataset = dataset.shuffle(shuffle_buffer_size)
+    dataset = dataset.repeat()
+
+  dataset = bin_and_batch(
+      dataset,
+      training,
+      n_devices,
+      batch_size=batch_size,
+      max_eval_length=max_eval_target_length,
+      drop_remainder=drop_remainder,
+      buckets=buckets)
+
+  if prefetch_size:
+    dataset = dataset.prefetch(prefetch_size)
+
+  return dataset
+
+
+def get_wmt_datasets(n_devices,
+                     dataset_name='wmt17_translate/de-en',
+                     data_dir=None,
```

Comment:
Ah that's just an oversight - I need to add the `data_dir` as a flag since people may want to specify a tensorflow_datasets directory other than `~/tensorflow_datasets` the default. fixing this.